File: src/pipeline/terrain-import/build-grid.ts
#!/usr/bin/env tsx

/**
 * Step 2: Merge downloaded GeoTIFF tiles into a single elevation raster.
 *
 * Uses GDAL's gdalwarp to merge and clip tiles to the region bounding box.
 * The output is a single GeoTIFF in the region's cache directory, which
 * extract-contours.ts reads directly.
 *
 * Requires: gdal (brew install gdal)
 */

import { execSync } from "child_process";
import { existsSync } from "fs";
import path from "path";
import {
  resolveRegion,
  loadRegionConfig,
  tilesDir,
  gridCacheDir,
} from "./util/region";
import { listLocalTiles } from "./util/grid-cache";

async function main(): Promise<void> {
  const slug = await resolveRegion(process.argv.slice(2));
  const config = loadRegionConfig(slug);
  const bbox = config.bbox;
  const tiles = tilesDir(slug);
  const cache = gridCacheDir(slug);

  console.log(`Region: ${config.name}`);
  console.log(
    `BBOX: ${bbox.minLat.toFixed(4)},${bbox.minLon.toFixed(4)} → ${bbox.maxLat.toFixed(4)},${bbox.maxLon.toFixed(4)}`,
  );

  const localTilePaths = listLocalTiles(tiles, bbox);
  if (localTilePaths.length === 0) {
    throw new Error(
      `No matching GeoTIFF files in ${tiles}. Run download step first.`,
    );
  }

  const outputPath = path.join(cache, "merged.tif");
  const force = process.argv.includes("--force");

  if (!force && existsSync(outputPath)) {
    console.log(`Merged grid already exists: ${outputPath}`);
    console.log("Use --force to rebuild.");
    return;
  }

  // Ensure cache directory exists
  execSync(`mkdir -p "${cache}"`);

  console.log(`Merging ${localTilePaths.length} tiles with gdalwarp...`);

  const t0 = performance.now();
  const tileArgs = localTilePaths.map((p) => `"${p}"`).join(" ");
  const cmd = [
    "gdalwarp",
    "-t_srs EPSG:4326",
    `-te ${bbox.minLon} ${bbox.minLat} ${bbox.maxLon} ${bbox.maxLat}`,
    "-overwrite",
    tileArgs,
    `"${outputPath}"`,
  ].join(" ");

  execSync(cmd, { stdio: "inherit" });

  const elapsed = performance.now() - t0;
  console.log(`Merged grid: ${outputPath}  (${elapsed.toFixed(0)}ms)`);
}

main().catch((error) => {
  console.error(error instanceof Error ? error.message : error);
  process.exit(1);
});

File: pipeline/build-level/src/build_grid.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use anyhow::{bail, Context, Result};
use pipeline_core::humanize::format_int;
use pipeline_core::step::{format_ms, StepView};

use crate::geo::{bbox_intersects, parse_tile_coverage_from_name};
use crate::region::{
    display_path, grid_cache_dir, load_region_config, resolve_region, tiles_dir, BoundingBox,
};

pub fn run_build_grid(region_arg: Option<&str>, force: bool, view: &StepView) -> Result<()> {
    let slug = resolve_region(region_arg)?;
    let config = load_region_config(&slug)?;

    let bbox = config.effective_bbox();
    view.info(format!("Region: {}", slug));
    view.info(format!(
        "BBOX: {:.4},{:.4} -> {:.4},{:.4}",
        bbox.min_lat, bbox.min_lon, bbox.max_lat, bbox.max_lon
    ));

    let local_tile_paths = list_local_tiles(&tiles_dir(&slug), &bbox)?;
    if local_tile_paths.is_empty() {
        bail!(
            "No matching GeoTIFF files in {}. Run download step first.",
            tiles_dir(&slug).display()
        );
    }

    let cache_dir = grid_cache_dir(&slug);
    let output_path = cache_dir.join("merged.tif");

    if !force && output_path.exists() {
        view.info(format!(
            "Merged grid already exists: {}",
            display_path(&output_path)
        ));
        view.info("Use --force to rebuild.");
        return Ok(());
    }

    fs::create_dir_all(&cache_dir)
        .with_context(|| format!("Failed to create {}", cache_dir.display()))?;

    let label = format!(
        "Merging {} tiles with gdalwarp",
        format_int(local_tile_paths.len())
    );
    view.try_run_step(
        &label,
        || {
            let mut cmd = Command::new("gdalwarp");
            cmd.arg("-t_srs")
                .arg("EPSG:4326")
                .arg("-te")
                .arg(bbox.min_lon.to_string())
                .arg(bbox.min_lat.to_string())
                .arg(bbox.max_lon.to_string())
                .arg(bbox.max_lat.to_string())
                .arg("-overwrite");

            for path in &local_tile_paths {
                cmd.arg(path);
            }
            cmd.arg(&output_path);

            let status = cmd
                .status()
                .context("Failed to execute gdalwarp. Install GDAL (brew install gdal)")?;

            if !status.success() {
                bail!("gdalwarp failed with status {status}");
            }
            Ok(())
        },
        |_, d| {
            format!(
                "Merged grid: {}  ({}ms)",
                display_path(&output_path),
                format_ms(d)
            )
        },
    )
}

pub fn list_local_tiles(tiles_dir: &Path, target_bbox: &BoundingBox) -> Result<Vec<PathBuf>> {
    if !tiles_dir.exists() {
        return Ok(Vec::new());
    }

    let mut paths = Vec::new();
    for entry in fs::read_dir(tiles_dir)
        .with_context(|| format!("Failed to read {}", tiles_dir.display()))?
    {
        let entry = entry?;
        let path = entry.path();

        let Some(name) = path.file_name().and_then(|s| s.to_str()) else {
            continue;
        };

        if !name.to_ascii_lowercase().ends_with(".tif") {
            continue;
        }

        let include = if let Some(coverage) = parse_tile_coverage_from_name(name) {
            bbox_intersects(&coverage, target_bbox)
        } else {
            true
        };

        if include {
            paths.push(path);
        }
    }

    paths.sort();
    Ok(paths)
}
